File: src/net/acmicpc/solved/silver/___Solution_11660.java
package net.acmicpc.solved.silver;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;

public class ___Solution_11660 {

	private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

	// 구간 합 구하기 5 - S1
	// https://www.acmicpc.net/problem/11660
	public static void main(String[] args) throws IOException {
		int[] input = readIntegers();
		int N = input[0];
		int M = input[1];

		int[][] prefixSumsArr = new int[N + 1][N + 1];
		for (int i = 1; i < prefixSumsArr.length; i++) {
			int[] numbers = readIntegers();

			for (int j = 1; j < prefixSumsArr[i].length; j++) {
				prefixSumsArr[i][j] = (numbers[j - 1] + prefixSumsArr[i - 1][j] + prefixSumsArr[i][j - 1]) - prefixSumsArr[i - 1][j - 1];
			}
		}

		// #logging
		//		for (int[] line : prefixSumsArr) {
		//			write(Arrays.toString(line));
		//		}

		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < M; i++) {
			int[] xy = readIntegers();
			int x1 = xy[0];
			int y1 = xy[1];
			int x2 = xy[2];
			int y2 = xy[3];

			int sum = (prefixSumsArr[x2][y2] - prefixSumsArr[x2][y1 - 1] - prefixSumsArr[x1 - 1][y2]) + prefixSumsArr[x1 - 1][y1 - 1];
			sb.append(sum).append("\n");
		}

		write(sb);
		close();
	}


	// Libraries
	private static String readLine() throws IOException {
		return br.readLine();
	}

	private static int readInt() throws IOException {
		return Integer.parseInt(br.readLine()
								  .trim());
	}

	private static int[] readIntegers() throws IOException {
		return Arrays.stream(br.readLine()
							   .trim()
							   .split(" "))
					 .mapToInt(Integer::parseInt)
					 .toArray();
	}

	private static <T> void write(T value) throws IOException {
		String s = String.valueOf(value);
		bw.write(s);
		bw.newLine();
	}

	private static void close() throws IOException {
		bw.flush();
		bw.close();
		br.close();
	}

}

File: src/net/acmicpc/solved/silver/Solution_11660.py
import sys

# libs
input = sys.stdin.readline


# 구간 합 구하기 - S1
# https://www.acmicpc.net/problem/11660

# Question
# N×N 크기의 표에 수가 채워져 있다.
# (x1, y1)부터 (x2, y2)까지 합을 구하는 프로그램을 작성
#

# 1. 입력받기
# N - 표의 크기
# M - 합을 구해야 하는 횟수
N, M = list(map(int, input().split()))

# 2. 표 입력 받기
matrix = []
for _ in range(N):
    row = list(map(int, input().split()))
    matrix.append(row)

# 3. 누적합 계산
prefix_sum = [[0] * (N + 1) for _ in range(N + 1)]
for i in range(1, N + 1):
    for j in range(1, N + 1):
        prefix_sum[i][j] = (
            matrix[i - 1][j - 1]  # 현재 위치핞 좌표의 값
            + prefix_sum[i - 1][j]  # 위쪽 누적합
            + prefix_sum[i][j - 1]  # 왼쪽 누적합
            - prefix_sum[i - 1][j - 1]  # 중복된 영역 제거(왼쪽 위)
        )

# 4. 결과 계산 & 출력
for _ in range(M):
    x1, y1, x2, y2 = map(int, input().split())
    result = prefix_sum[x2][y2] - prefix_sum[x1 - 1][y2] - prefix_sum[x2][y1 - 1] + prefix_sum[x1 - 1][y1 - 1]
    print(result)
